vaultclient/vaultclient.go

```diff
@@ -1,63 +1,175 @@
+				secretKey,
+				sessionToken,
+			),
+		},
+		client:   &http.Client{},
+		endpoint: endpoint,
+	}
+}
+
+// makeAWSRequest is a helper function that handles AWS v4 signing and HTTP requests
+// This centralizes the common request logic used by all API operations
+func (c *Vault) makeAWSRequest(ctx context.Context, action string, formData url.Values) (*http.Response, error) {
+	formData.Set("Action", action)
+	formData.Set("Version", "2010-05-08")
+	endpoint := c.endpoint + "/"
+
+	requestBody := formData.Encode()
+	req, err := http.NewRequestWithContext(ctx, "POST", endpoint, strings.NewReader(requestBody))
+	if err != nil {
+		return nil, fmt.Errorf("failed to create request: %w", err)
+	}
+
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+	req.Header.Set("Host", req.URL.Host)
+
+	// Hash the request body payload for AWS signing
+	hash := sha256.Sum256([]byte(requestBody))
+	payloadHash := hex.EncodeToString(hash[:])
+
+	// Sign the request using AWS SDK v2 signer
+	signer := v4.NewSigner()
+	creds, err := c.cfg.Credentials.Retrieve(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve credentials: %w", err)
 	}
 
-	// Handlers
-	svc.Handlers.Sign.PushBackNamed(v4.SignRequestHandler)
-	svc.Handlers.Build.PushBackNamed(query.BuildHandler)
-	svc.Handlers.Unmarshal.PushBackNamed(restjson.UnmarshalHandler)
-	svc.Handlers.UnmarshalError.PushBackNamed(query.UnmarshalErrorHandler)
+	err = signer.SignHTTP(ctx, creds, req, payloadHash, ServiceName, c.cfg.Region, time.Now())
+	if err != nil {
+		return nil, fmt.Errorf("failed to sign request: %w", err)
+	}
 
-	return svc
+	resp, err := c.client.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("failed to make request: %w", err)
+	}
+
+	return resp, nil
 }
 
-// newRequest creates a new request for a Vault operation and runs any
-// custom request initialization.
-func (c *Vault) newRequest(op *request.Operation, params, data interface{}) *request.Request {
-	req := c.NewRequest(op, params, data)
+type APIError struct {
+	StatusCode int
+	Err        error
+}
+
+func (h *APIError) Error() string {
+	return h.Err.Error()
+}
+
+func (h *APIError) Unwrap() error {
+	return h.Err
+}
+
+func (h *APIError) HTTPStatusCode() int {
+	return h.StatusCode
+}
+
+func (vc *Vault) handleAWSResponse(resp *http.Response) (respBody []byte, err error) {
+	defer func() {
+		if closeErr := resp.Body.Close(); closeErr != nil && err == nil {
+			err = fmt.Errorf("failed to close response body: %w", closeErr)
+		}
+	}()
+
+	respBody, err = io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read response: %w", err)
+	}
+
+	if resp.StatusCode >= 400 {
+		var xmlError struct {
+			Error struct {
+				Code    string `xml:"Code"`
+				Message string `xml:"Message"`
+			} `xml:"Error"`
+			RequestID string `xml:"RequestId"`
+		}
+		if err := xml.Unmarshal(respBody, &xmlError); err == nil {
+			switch xmlError.Error.Code {
+			case "EntityAlreadyExists":
```

Comment:
can we retrieve the error code from the type?

```suggestion
			case types.EntityAlreadyExistsException{}.ErrorCode():
```

Reply:
Yes it works